### ISSUE TITLE
Exception when sending embeds on multiple servers

### ISSUE THREAD

**Issue by NotOfficer** [RogueException/Discord.Net#1374]
As mentioned in the title, this happens when sending embeds on multiple servers at once.
The installed version is the latest nightly build (2.2.0-dev-20190908.10)

Exception:
```
System.ArgumentOutOfRangeException: Valid values are between -62135596800000 and 253402300799999, inclusive.
Parameter name: milliseconds
   at System.DateTimeOffset.FromUnixTimeMilliseconds(Int64 milliseconds)
   at Discord.Net.RateLimitInfo..ctor(Dictionary`2 headers)
   at Discord.Net.Queue.RequestBucket.SendAsync(RestRequest request)
   at Discord.Net.Queue.RequestQueue.SendAsync(RestRequest request)
   at Discord.API.DiscordRestApiClient.SendInternalAsync(String method, String endpoint, RestRequest request)
   at Discord.API.DiscordRestApiClient.SendJsonAsync[TResponse](String method, String endpoint, Object payload, String bucketId, ClientBucketType clientBucket, RequestOptions options)
   at Discord.API.DiscordRestApiClient.CreateMessageAsync(UInt64 channelId, CreateMessageParams args, RequestOptions options)
   at Discord.Rest.ChannelHelper.SendMessageAsync(IMessageChannel channel, BaseDiscordClient client, String text, Boolean isTTS, Embed embed, RequestOptions options)```

**Comment by k-boyle** [comment]
Looks like #1354 broke something

**Comment by Chris-Johnston** [comment]
~~As a temporary fix, does setting `RateLimitPrecision` to seconds in your client config resolve this issue?~~
Edit: Actually, I don't think that would fix it, the logic in that area is the same independent of that setting.

**Comment by Chris-Johnston** [comment]
I'm having trouble reproducing this error. I'll try later to hit rate limits using the same method you mentioned. Is it consistent?
Odd that the API is sending you that value. Maybe it's parsing the header value as `NaN` or Infinity, but I would expect a different error from that. Even without the latest changes I'd expect the same exception for the header value you received. Maybe this can be expected from the API?
I've double-checked that the `X-Reset-After` header isn't actually sending milliseconds when precision is set to milliseconds, but it seems that it's acting as expected by the docs.

**Comment by NotOfficer** [comment]
Im currently not at home. When im back i may create a project where the issue is reproducable. But i remember it was throwing 3/3 times then i went back to the stable version without any issue.

**Comment by NotOfficer** [comment]
So I've made a simple bot and it appears to even throw if i just send a single message:

```cs
private static DiscordSocketClient _client;

private static void Main() => RunBot().GetAwaiter().GetResult();

private static async Task RunBot()
{
	_client = new DiscordSocketClient(new DiscordSocketConfig
	{
		LogLevel = LogSeverity.Debug
	});

	_client.Log += Client_Log;
	_client.Ready += Client_Ready;

	await _client.LoginAsync(TokenType.Bot, _token).ConfigureAwait(false);
	await _client.StartAsync().ConfigureAwait(false);
	await Task.Delay(-1).ConfigureAwait(false);
}

private static Task Client_Ready()
{
	var guild = _client.GetGuild(_guildId);
	var channel = guild.GetTextChannel(_testChannelId);
	return channel.SendMessageAsync("Client ready");
}

private static async Task Client_Log(LogMessage arg)
{
	Console.WriteLine($"[{DateTime.Now} - {arg.Source}] {arg.Message}");

	if (arg.Exception != null && _client.ConnectionState == ConnectionState.Connected)
	{
		var guild = _client.GetGuild(_guildId);
		var channel = guild.GetTextChannel(_testChannelId);

		using (var ms = new MemoryStream(Encoding.UTF8.GetBytes(arg.Exception.ToString())))
		{
			await channel.SendFileAsync(ms, $"{DateTime.UtcNow:yyyy_MM_dd-HH_mm_ss}.txt", arg.Exception.Message).ConfigureAwait(false);
		}
	}
}
```

Exception:
https://cdn.discordapp.com/attachments/622161732491804706/622170501162336256/2019_09_13-20_43_25.txt

Channel output:
![](https://media.wtf/17473611.png)

Console output:
```
[13.09.2019 20:43:22 - Discord] Discord.Net v2.2.0-dev-20190908.10 (API v6)
[13.09.2019 20:43:23 - Rest] GET voice/regions: 389,45 ms
[13.09.2019 20:43:23 - Rest] GET users/@me: 128,89 ms
[13.09.2019 20:43:23 - Gateway] Connecting
[13.09.2019 20:43:23 - Gateway] Connecting ApiClient
[13.09.2019 20:43:23 - Rest] GET gateway: 14,24 ms
[13.09.2019 20:43:24 - Gateway] Received Hello
[13.09.2019 20:43:24 - Gateway] Heartbeat Started
[13.09.2019 20:43:24 - Gateway] Identifying
[13.09.2019 20:43:24 - Gateway] Sent Heartbeat
[13.09.2019 20:43:24 - Gateway] Sent Identify
[13.09.2019 20:43:24 - Gateway] Received HeartbeatAck
[13.09.2019 20:43:24 - Gateway] Latency = 156 ms
[13.09.2019 20:43:24 - Gateway] Received Dispatch (READY)
[13.09.2019 20:43:24 - Gateway] GuildDownloader Started
[13.09.2019 20:43:24 - Gateway] Sending Status
[13.09.2019 20:43:24 - Gateway] Sent StatusUpdate
[13.09.2019 20:43:24 - Gateway] Connected
[13.09.2019 20:43:24 - Gateway] Raising Event
[13.09.2019 20:43:24 - Gateway] Received Dispatch (GUILD_AVAILABLE)
[13.09.2019 20:43:24 - Gateway] Connected to Rate Limit Test
[13.09.2019 20:43:24 - Gateway] GuildDownloader Stopped
[13.09.2019 20:43:25 - Gateway] Received Dispatch (MESSAGE_CREATE)
[13.09.2019 20:43:25 - Gateway] A Ready handler has thrown an unhandled exception.
[13.09.2019 20:43:25 - Gateway] Received Dispatch (MESSAGE_CREATE)
[13.09.2019 20:43:25 - Gateway] Ready
```

**Comment by NotOfficer** [comment]
> As a temporary fix, does setting `RateLimitPrecision` to seconds in your client config resolve this issue?

With that setting it doesnt throw.

**Comment by Chris-Johnston** [comment]
wat. Well, at least that narrows it down. Thanks for the help debugging this.

**Comment by Chris-Johnston** [comment]
Works on my machine™️ , no exceptions here when running the same program.
![image](https://user-images.githubusercontent.com/16418643/64902380-e4784a00-d65a-11e9-8fe4-73c4b6c999bc.png)

If you are able, try cloning the repo and directly linking to the projects in it. [Placing a breakpoint here would](https://github.com/discord-net/Discord.Net/blob/9482204bcfdc57c859f4fd6daae0c44d3010e931/src/Discord.Net.Rest/Net/RateLimitInfo.cs#L16) to look at the `X-RateLimit-Reset` header would help to figure out why it's so far out of range.

Also, I found that a float isn't precise enough to store the unix time w/ milliseconds, but I wouldn't expect that to make the range so far off (I saw it would be a little bit behind). #1375 currently addresses this, but I don't think we publish build artifacts for pull requests anymore so there's no quick way to try it out without pulling my fork's branch.
